lib/HLSL/DxilPreparePasses.cpp

```diff
@@ -640,6 +640,83 @@ class DxilFinalizeModule : public ModulePass {
     }
   }
 
+  void convertIsInf(Module &M, CallInst *CI) {
+    IRBuilder<> Builder(CI);
+    Value *Val = CI->getOperand(1);
+    Type *IType = Type::getInt16Ty(M.getContext());
+    Constant *Inf1 = ConstantInt::get(IType, 0x7c00);
+    Constant *Inf2 = ConstantInt::get(IType, 0xfc00);
+
+    Value *IVal = Builder.CreateBitCast(Val, IType);
+    Value *B1 = Builder.CreateICmpEQ(IVal, Inf1);
+    Value *B2 = Builder.CreateICmpEQ(IVal, Inf2);
+    Value *B3 = Builder.CreateOr(B1, B2);
+    CI->replaceAllUsesWith(B3);
+    CI->eraseFromParent();
+  }
+
+  void convertIsNaN(Module &M, CallInst *CI) {
+    IRBuilder<> Builder(CI);
+    Value *Val = CI->getOperand(1);
+    Type *IType = Type::getInt16Ty(M.getContext());
+
+    Constant *ExpBitMask = ConstantInt::get(IType, 0x7c00);
+    Constant *SigBitMask = ConstantInt::get(IType, 0x3ff);
+
+    Value *IVal = Builder.CreateBitCast(Val, IType);
+    Value *Exp = Builder.CreateAnd(IVal, ExpBitMask);
+    Value *B1 = Builder.CreateICmpEQ(Exp, ExpBitMask);
+
+    Value *Sig = Builder.CreateAnd(IVal, SigBitMask);
+    Value *B2 = Builder.CreateICmpNE(Sig, ConstantInt::get(IType, 0));
+    Value *B3 = Builder.CreateAnd(B1, B2);
+    CI->replaceAllUsesWith(B3);
+    CI->eraseFromParent();
+  }
+
+  void convertIsFinite(Module &M, CallInst *CI) {
+    IRBuilder<> Builder(CI);
+    Value *Val = CI->getOperand(1);
+    Type *IType = Type::getInt16Ty(M.getContext());
+
+    Constant *ExpBitMask = ConstantInt::get(IType, 0x7c00);
+
+    Value *IVal = Builder.CreateBitCast(Val, IType);
+    Value *Exp = Builder.CreateAnd(IVal, ExpBitMask);
+    Value *B1 = Builder.CreateICmpNE(Exp, ExpBitMask);
+    CI->replaceAllUsesWith(B1);
+    CI->eraseFromParent();
+  }
+
+  void convertIsSpecialFloat(Module &M, hlsl::OP *hlslOP) {
+    for (auto FnIt : hlslOP->GetOpFuncList(DXIL::OpCode::IsInf)) {
```

Comment:
Is this the right opcode to loop over? I would have thought it would be something like OpCode::IsSpecialFloat.
Why IsInf? Couldn't it be IsNan or IsFinite too?

Reply:
I didn't see a  OpCode::IsSpecialFloat; yes isnan or isfinite could have been used as well.

Reply:
This might be worth explaining in a comment.  Naive reader: this looks for IsInf's, but it seems that this is doing more than that.